File: src/examples/py/pb_chn_bench/benchmark_publisher_module.py
# Copyright (c) 2024 The AimRT Authors.
# AimRT is licensed under Mulan PSL v2.

import random
import string
import threading
import time

import aimrt_py
import benchmark_pb2
import yaml


class BenchmarkPublisher(aimrt_py.ModuleBase):
    def __init__(self):
        super().__init__()
        self.core = aimrt_py.CoreRef()
        self.logger = aimrt_py.LoggerRef()
        self.max_topic_number = 0
        self.max_parallel_number = 0
        self.executor_vec = []
        self.publisher_vec = []
        self.bench_plans = []
        self.run_flag = True
        self.stop_sig = threading.Event()
        self.shutdown_delay = 1  # seconds
        self.bench_interval = 1  # seconds
        self.publish_complete_event = threading.Event()

    def Info(self) -> aimrt_py.ModuleInfo:
        info = aimrt_py.ModuleInfo()
        info.name = "BenchmarkPublisherModule"
        return info

    def Initialize(self, core: aimrt_py.CoreRef) -> bool:
        assert isinstance(core, aimrt_py.CoreRef)

        self.core = core
        self.logger = self.core.GetLogger()

        aimrt_py.info(self.logger, "Module initializing ...")

        try:
            # read config
            file_path = self.core.GetConfigurator().GetConfigFilePath()
            with open(file_path, "r") as f:
                cfg_node = yaml.safe_load(f)

            self.max_topic_number = cfg_node.get("max_topic_number", 0)
            self.max_parallel_number = cfg_node.get("max_parallel_number", 0)

            if "bench_plans" in cfg_node and isinstance(cfg_node["bench_plans"], list):
                for bench_plan_node in cfg_node["bench_plans"]:
                    print(f"bench_plan_node: {bench_plan_node}")
                    if "perf_mode" in bench_plan_node:
                        perf_mode = bench_plan_node["perf_mode"]
                    else:
                        raise ValueError("Bench plan perf_mode is not set")

                    if perf_mode == "multi-topic":
                        bench_plan = {
                            "perf_mode": perf_mode,
                            "channel_frq": bench_plan_node["channel_frq"],
                            "msg_size": bench_plan_node["msg_size"],
                            "topic_number": bench_plan_node["topic_number"],
                            "msg_count": bench_plan_node["msg_count"],
                        }
                        if bench_plan["topic_number"] > self.max_topic_number:
                            raise ValueError(
                                f"Bench plan topic number ({bench_plan['topic_number']}) "
                                f"is greater than max topic number ({self.max_topic_number})"
                            )
                    elif perf_mode == "parallel":
                        bench_plan = {
                            "perf_mode": perf_mode,
                            "channel_frq": bench_plan_node["channel_frq"],
                            "msg_size": bench_plan_node["msg_size"],
                            "parallel_number": bench_plan_node["parallel_number"],
                            "msg_count": bench_plan_node["msg_count"],
                        }
                    else:
                        raise ValueError("Unsupport perf mode: " + perf_mode)

                    self.bench_plans.append(bench_plan)

            aimrt_py.info(
                self.logger,
                f"Module config: max_topic_number={self.max_topic_number}, " f"bench_plans={self.bench_plans}",
            )

            # controller
            self.publish_control_executor = self.core.GetExecutorManager().GetExecutor("publish_control_executor")
            if not self.publish_control_executor:
                raise RuntimeError("Get executor 'publish_control_executor' failed.")

            # signal publisher
            signal_topic = "benchmark_signal"
            self.signal_publisher = self.core.GetChannelHandle().GetPublisher(signal_topic)
            if not self.signal_publisher:
                raise RuntimeError(f"Get publisher for topic '{signal_topic}' failed.")

            ret = aimrt_py.RegisterPublishType(self.signal_publisher, benchmark_pb2.BenchmarkSignal)
            if not ret:
                raise RuntimeError(f"Register publish type for topic '{signal_topic}' failed.")

            # message publishers
            for ii in range(max(self.max_topic_number, self.max_parallel_number)):
                executor_name = f"publish_executor_{ii}"
                executor = self.core.GetExecutorManager().GetExecutor(executor_name)
                if not executor:
                    raise RuntimeError(f"Get executor '{executor_name}' failed.")

                self.executor_vec.append(executor)

            for ii in range(self.max_topic_number):
                topic_name = f"test_topic_{ii}"
                publisher = self.core.GetChannelHandle().GetPublisher(topic_name)
                if not publisher:
                    raise RuntimeError(f"Get publisher for topic '{topic_name}' failed.")

                ret = aimrt_py.RegisterPublishType(publisher, benchmark_pb2.BenchmarkMessage)
                if not ret:
                    raise RuntimeError(f"Register publish type for topic '{topic_name}' failed.")

                self.publisher_vec.append(publisher)

        except Exception as e:
            aimrt_py.error(self.logger, f"Initialize failed: {e}")
            return False

        aimrt_py.info(self.logger, "Init succeeded")
        return True

    def Start(self) -> bool:
        try:
            aimrt_py.info(self.logger, "Module starting ...")
            self.publish_control_executor.Execute(self.MainLoop)

        except Exception as e:
            aimrt_py.error(self.logger, f"Start failed: {e}")
            return False

        aimrt_py.info(self.logger, "Start succeeded")
        return True

    def Shutdown(self) -> bool:
        try:
            aimrt_py.info(self.logger, "Module shutting down ...")

            self.run_flag = False
            self.publish_complete_event.set()
            self.stop_sig.wait()
        except Exception as e:
            aimrt_py.error(self.logger, f"Shutdown failed: {e}")
            return False

        aimrt_py.info(self.logger, "Shutdown succeeded")
        return True

    def MainLoop(self) -> None:
        try:
            aimrt_py.info(self.logger, "Start Bench.")

            # warm up
            for ii in range(10):
                begin_signal = benchmark_pb2.BenchmarkSignal()
                begin_signal.status = benchmark_pb2.BenchmarkStatus.WarmUp
                aimrt_py.Publish(self.signal_publisher, begin_signal)
                time.sleep(0.1)
                aimrt_py.info(self.logger, f"Warm up {ii}")

            # benchmark
            for ii, bench_plan in enumerate(self.bench_plans):
                if not self.run_flag:
                    break

                print(f"Start bench plan {ii}")
                self.StartSinglePlan(ii, bench_plan)
                print(f"End bench plan {ii}")

                time.sleep(self.bench_interval)

        except Exception as e:
            aimrt_py.error(self.logger, f"Exit MainLoop with exception: {e}")

        self.stop_sig.set()

    def StartSinglePlan(self, plan_id: int, plan: dict) -> None:
        if plan["perf_mode"] == "multi-topic":
            self.StartMultiTopicPlan(plan_id, plan)
        else:
            self.StartParallelPlan(plan_id, plan)

    def StartMultiTopicPlan(self, plan_id: int, plan: dict) -> None:
        # publish start signal
        begin_signal = benchmark_pb2.BenchmarkSignal()
        begin_signal.status = benchmark_pb2.BenchmarkStatus.Begin
        begin_signal.mode = "multi-topic"
        begin_signal.bench_plan_id = plan_id
        begin_signal.topic_number = plan['topic_number']
        begin_signal.parallel_number = 1
        begin_signal.send_num = plan['msg_count']
        begin_signal.message_size = plan['msg_size']
        begin_signal.send_frequency = plan['channel_frq']

        aimrt_py.info(self.logger, f"Publish benchmark start signal, data: \n{begin_signal}")
        aimrt_py.Publish(self.signal_publisher, begin_signal)

        self.publish_complete_event.clear()
        self.completed_tasks = 0
        self.total_tasks = plan['topic_number']

        time.sleep(1)

        # publish topic
        for ii in range(plan['topic_number']):
            publish_executor = self.executor_vec[ii]
            publish_executor.Execute(lambda index=ii: self.PublishTask(self.publisher_vec[index], plan))

        self.publish_complete_event.wait()

        # wait for subscriber to receive all messages
        time.sleep(1)

        # publish end signal
        end_signal = benchmark_pb2.BenchmarkSignal()
        end_signal.status = benchmark_pb2.BenchmarkStatus.End
        end_signal.bench_plan_id = plan_id
        aimrt_py.info(self.logger, f"Publish benchmark end signal, data: {end_signal}")
        aimrt_py.Publish(self.signal_publisher, end_signal)

    def StartParallelPlan(self, plan_id: int, plan: dict) -> None:
        # publish start signal
        begin_signal = benchmark_pb2.BenchmarkSignal()
        begin_signal.status = benchmark_pb2.BenchmarkStatus.Begin
        begin_signal.mode = "parallel"
        begin_signal.bench_plan_id = plan_id
        begin_signal.topic_number = 1
        begin_signal.parallel_number = plan["parallel_number"]
        begin_signal.send_num = plan["msg_count"]
        begin_signal.message_size = plan["msg_size"]
        begin_signal.send_frequency = plan["channel_frq"]

        aimrt_py.info(self.logger, f"Publish benchmark start signal, data: \n{begin_signal}")
        aimrt_py.Publish(self.signal_publisher, begin_signal)

        time.sleep(1)

        # Generate last send counts for message distribution
        last_send_counts = self.generate_last_send_counts(plan["msg_count"], plan["parallel_number"])

        self.publish_complete_event.clear()
        self.completed_tasks = 0
        self.total_tasks = plan["parallel_number"]

        # publish topic
        for ii in range(plan["parallel_number"]):
            publish_executor = self.executor_vec[ii]
            publisher = self.publisher_vec[0]  # Always use the first publisher
            messages_to_send = last_send_counts[ii + 1] - last_send_counts[ii]

            publish_executor.Execute(
                lambda index=ii,
                msgs=messages_to_send,
                start_seq=last_send_counts[ii],
                pub=publisher: self.PublishParallelTask(pub, plan, msgs, start_seq)
            )

        # Wait for all tasks to complete
        self.publish_complete_event.wait()

        time.sleep(1)

        # publish end signal
        end_signal = benchmark_pb2.BenchmarkSignal()
        end_signal.status = benchmark_pb2.BenchmarkStatus.End
        end_signal.bench_plan_id = plan_id
        aimrt_py.info(self.logger, f"Publish benchmark end signal, data: {end_signal}")
        aimrt_py.Publish(self.signal_publisher, end_signal)

    def PublishParallelTask(self, publisher, plan: dict, messages_to_send: int, start_seq: int) -> None:
        msg = benchmark_pb2.BenchmarkMessage()
        msg.data = self.GenerateRandomString(plan["msg_size"])

        send_count = 0
        sleep_ns = 1_000_000_000 / plan["channel_frq"]  # Convert to nanoseconds

        # Use a more precise timing approach
        next_time = time.time_ns()

        while send_count < messages_to_send and self.run_flag:
            msg.seq = start_seq + send_count
            msg.timestamp = time.time_ns()
            aimrt_py.Publish(publisher, msg)

            send_count += 1
            next_time += sleep_ns

            # Sleep until next scheduled time
            current_time = time.time_ns()
            if next_time > current_time:
                time.sleep((next_time - current_time) / 1_000_000_000)  # Convert back to seconds

        with threading.Lock():
            self.completed_tasks += 1
            if self.completed_tasks == self.total_tasks:
                self.publish_complete_event.set()

    @staticmethod
    def generate_last_send_counts(msg_count: int, parallel_number: int) -> list:
        """Generate the last send counts for each parallel task."""
        result = [0]
        base_count = msg_count // parallel_number
        remainder = msg_count % parallel_number

        for i in range(parallel_number):
            extra = 1 if i < remainder else 0
            result.append(result[-1] + base_count + extra)

        return result

    def PublishTask(self, publisher: aimrt_py.PublisherRef, plan: dict) -> None:
        msg = benchmark_pb2.BenchmarkMessage()
        msg.data = self.GenerateRandomString(plan['msg_size'])

        send_count = 0
        sleep_time = 1 / plan['channel_frq']

        while send_count < plan['msg_count'] and self.run_flag:
            msg.seq = send_count
            msg.timestamp = time.time_ns()
            aimrt_py.Publish(publisher, msg)

            send_count += 1
            time.sleep(sleep_time)

        with threading.Lock():
            self.completed_tasks += 1
            if self.completed_tasks == self.total_tasks:
                self.publish_complete_event.set()

    @staticmethod
    def GenerateRandomString(length: int) -> bytes:
        return ''.join(random.choices(string.ascii_letters + string.digits, k=length)).encode()

File: src/examples/cpp/pb_chn/module/benchmark_publisher_module/benchmark_publisher_module.cc
// Copyright (c) 2023, AgiBot Inc.
// All rights reserved.

#include "benchmark_publisher_module/benchmark_publisher_module.h"
#include "aimrt_module_protobuf_interface/channel/protobuf_channel.h"
#include "aimrt_module_protobuf_interface/util/protobuf_tools.h"
#include "util/time_util.h"

#include "yaml-cpp/yaml.h"

#include "benchmark.pb.h"

namespace aimrt::examples::cpp::pb_chn::benchmark_publisher_module {

std::string GenerateRandomString(int min_length, int max_length) {
  static constexpr std::string_view kChars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
  srand(time(nullptr));

  int length = rand() % (max_length - min_length + 1) + min_length;

  std::string result;
  result.reserve(length);

  for (int i = 0; i < length; ++i) {
    result += kChars[rand() % kChars.length()];
  }

  return result;
}

std::string GenerateRandomString(int length) {
  return GenerateRandomString(length, length);
}

std::vector<uint32_t> GenerateLastSendCounts(uint32_t total, uint32_t concurrency) {
  if (concurrency <= 0) [[unlikely]]
    return {};
  if (concurrency == 1) return {0, total};

  std::vector<uint32_t> last_send_counts(concurrency + 1, 0);
  uint32_t base_count = total / concurrency;

  for (uint32_t i = 1; i < concurrency; ++i) {
    last_send_counts[i] = last_send_counts[i - 1] + base_count;
  }

  last_send_counts[concurrency] = total;

  return last_send_counts;
}

bool BenchmarkPublisherModule::Initialize(aimrt::CoreRef core) {
  core_ = core;

  try {
    // Read cfg
    auto file_path = core_.GetConfigurator().GetConfigFilePath();
    if (!file_path.empty()) {
      YAML::Node cfg_node = YAML::LoadFile(std::string(file_path));
      max_topic_number_ = cfg_node["max_topic_number"].as<uint32_t>();
      max_parallel_number_ = cfg_node["max_parallel_number"].as<uint32_t>();

      if (cfg_node["bench_plans"] && cfg_node["bench_plans"].IsSequence()) {
        for (const auto& bench_plan_node : cfg_node["bench_plans"]) {
          BenchPlan bench_plan;

          auto perf_mode = bench_plan_node["perf_mode"].as<std::string>();
          if (perf_mode == "multi-topic") {
            bench_plan.mode = BenchPlan::PerfMod::kMultiTopic;
          } else if (perf_mode == "parallel") {
            bench_plan.mode = BenchPlan::PerfMod::kParallel;
          } else {
            throw aimrt::common::util::AimRTException("Unsupport perf mode: " + perf_mode);
          }

          if (bench_plan.mode == BenchPlan::PerfMod::kParallel) {
            bench_plan.parallel_number = bench_plan_node["parallel_number"].as<uint32_t>();
            AIMRT_CHECK_ERROR_THROW(
                bench_plan.parallel_number <= max_parallel_number_,
                "Bench plan parallel number({}) is greater than max parallel number({})",
                bench_plan.parallel_number, max_parallel_number_);
          }

          if (bench_plan.mode == BenchPlan::PerfMod::kMultiTopic) {
            bench_plan.topic_number = bench_plan_node["topic_number"].as<uint32_t>();
            AIMRT_CHECK_ERROR_THROW(
                bench_plan.topic_number <= max_topic_number_,
                "Bench plan topic number({}) is greater than max topic number({})",
                bench_plan.topic_number, max_topic_number_);
          }

          bench_plan.channel_frq = bench_plan_node["channel_frq"].as<uint32_t>();
          bench_plan.msg_size = bench_plan_node["msg_size"].as<uint32_t>();
          bench_plan.msg_count = bench_plan_node["msg_count"].as<uint32_t>();

          bench_plans_.emplace_back(bench_plan);
        }
      }
    }

    // controller
    publish_control_executor_ = core_.GetExecutorManager().GetExecutor("publish_control_executor");
    AIMRT_CHECK_ERROR_THROW(publish_control_executor_, "Get executor 'publish_control_executor' failed.");

    signal_publisher_ = core_.GetChannelHandle().GetPublisher("benchmark_signal");
    AIMRT_CHECK_ERROR_THROW(signal_publisher_, "Get publisher for topic 'benchmark_signal' failed.");

    bool ret = aimrt::channel::RegisterPublishType<aimrt::protocols::example::BenchmarkSignal>(signal_publisher_);
    AIMRT_CHECK_ERROR_THROW(ret, "Register publish type failed.");

    for (uint32_t ii = 0; ii < std::max(max_parallel_number_, max_topic_number_); ++ii) {
      auto executor_name = "publish_executor_" + std::to_string(ii);
      auto executor = core_.GetExecutorManager().GetExecutor(executor_name);
      AIMRT_CHECK_ERROR_THROW(executor, "Get executor '{}' failed.", executor_name);

      executor_vec_.emplace_back(executor);
    }

    for (uint32_t ii = 0; ii < max_topic_number_; ++ii) {
      auto topic_name = "test_topic_" + std::to_string(ii);
      auto publisher = core_.GetChannelHandle().GetPublisher(topic_name);
      AIMRT_CHECK_ERROR_THROW(publisher, "Get publisher for topic '{}' failed.", topic_name);

      bool ret = aimrt::channel::RegisterPublishType<aimrt::protocols::example::BenchmarkMessage>(publisher);
      AIMRT_CHECK_ERROR_THROW(ret, "Register publish type failed.");

      publisher_vec_.emplace_back(publisher);
    }

  } catch (const std::exception& e) {
    AIMRT_ERROR("Init failed, {}", e.what());
    return false;
  }

  AIMRT_INFO("Init succeeded.");

  return true;
}

bool BenchmarkPublisherModule::Start() {
  try {
    publish_control_executor_.Execute(std::bind(&BenchmarkPublisherModule::MainLoop, this));
  } catch (const std::exception& e) {
    AIMRT_ERROR("Start failed, {}", e.what());
    return false;
  }

  AIMRT_INFO("Start succeeded.");
  return true;
}

void BenchmarkPublisherModule::Shutdown() {
  try {
    run_flag_ = false;

    stop_sig_.get_future().wait();

  } catch (const std::exception& e) {
    AIMRT_ERROR("Shutdown failed, {}", e.what());
    return;
  }

  AIMRT_INFO("Shutdown succeeded.");
}

// Main loop
void BenchmarkPublisherModule::MainLoop() {
  try {
    AIMRT_INFO("Start Bench.");

    // warm up
    for (size_t ii = 0; ii < 10; ++ii) {
      aimrt::protocols::example::BenchmarkSignal begin_signal;
      begin_signal.set_status(aimrt::protocols::example::BenchmarkStatus::WarmUp);
      aimrt::channel::Publish(signal_publisher_, begin_signal);
      std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    for (size_t ii = 0; ii < bench_plans_.size(); ++ii) {
      if (!run_flag_.load()) break;

      StartSinglePlan(ii, bench_plans_[ii]);

      std::this_thread::sleep_for(std::chrono::seconds(1));
    }

    AIMRT_INFO("Bench completed.");

  } catch (const std::exception& e) {
    AIMRT_ERROR("Exit MainLoop with exception, {}", e.what());
  }

  stop_sig_.set_value();
}

void BenchmarkPublisherModule::StartSinglePlan(uint32_t plan_id, BenchPlan plan) {
  if (plan.mode == BenchPlan::PerfMod::kMultiTopic) {
    StartMultiTopicPlan(plan_id, plan);
  } else {
    StartParallelPlan(plan_id, plan);
  }
}

void BenchmarkPublisherModule::StartMultiTopicPlan(uint32_t plan_id, BenchPlan plan) {
  // publish start signal
  aimrt::protocols::example::BenchmarkSignal begin_signal;
  begin_signal.set_status(aimrt::protocols::example::BenchmarkStatus::Begin);
  begin_signal.set_mode("multi-topic");
  begin_signal.set_bench_plan_id(plan_id);
  begin_signal.set_topic_number(plan.topic_number);
  begin_signal.set_parallel_number(1);
  begin_signal.set_send_num(plan.msg_count);
  begin_signal.set_message_size(plan.msg_size);
  begin_signal.set_send_frequency(plan.channel_frq);

  AIMRT_INFO("Publish benchmark start signal, data: {}", aimrt::Pb2CompactJson(begin_signal));
  aimrt::channel::Publish(signal_publisher_, begin_signal);

  std::this_thread::sleep_for(std::chrono::seconds(1));

  // publish topic
  std::vector<std::future<void>> future_vec;
  for (size_t ii = 0; ii < plan.topic_number; ++ii) {
    auto publish_executor = executor_vec_[ii];
    auto publisher = publisher_vec_[ii];

    std::promise<void> task_promise;
    future_vec.emplace_back(task_promise.get_future());
    publish_executor.Execute(
        [this, publisher, plan, task_promise{std::move(task_promise)}]() mutable {
          aimrt::protocols::example::BenchmarkMessage msg;
          msg.set_data(GenerateRandomString(plan.msg_size));

          uint32_t send_count = 0;

          uint32_t sleep_ns = static_cast<uint32_t>(1000000000 / plan.channel_frq);
          auto cur_tp = std::chrono::steady_clock::now();

          for (; send_count < plan.msg_count; ++send_count) {
            if (!run_flag_.load()) [[unlikely]]
              break;

            msg.set_seq(send_count);
            msg.set_timestamp(aimrt::common::util::GetCurTimestampNs());

            aimrt::channel::Publish(publisher, msg);

            cur_tp += std::chrono::nanoseconds(sleep_ns);
            std::this_thread::sleep_until(cur_tp);
          }

          task_promise.set_value();
        });
  }

  for (auto& fu : future_vec) fu.wait();

  std::this_thread::sleep_for(std::chrono::seconds(1));

  // publish end signal
  aimrt::protocols::example::BenchmarkSignal end_signal;
  end_signal.set_status(aimrt::protocols::example::BenchmarkStatus::End);
  end_signal.set_bench_plan_id(plan_id);

  AIMRT_INFO("Publish benchmark end signal, data: {}", aimrt::Pb2CompactJson(end_signal));
  aimrt::channel::Publish(signal_publisher_, end_signal);
}

void BenchmarkPublisherModule::StartParallelPlan(uint32_t plan_id, BenchPlan plan) {
  // publish start signal
  aimrt::protocols::example::BenchmarkSignal begin_signal;
  begin_signal.set_status(aimrt::protocols::example::BenchmarkStatus::Begin);
  begin_signal.set_bench_plan_id(plan_id);
  begin_signal.set_mode("parallel");
  begin_signal.set_topic_number(1);
  begin_signal.set_parallel_number(plan.parallel_number);
  begin_signal.set_send_num(plan.msg_count);
  begin_signal.set_message_size(plan.msg_size);
  begin_signal.set_send_frequency(plan.channel_frq);

  AIMRT_INFO("Publish benchmark start signal, data: {}", aimrt::Pb2CompactJson(begin_signal));
  aimrt::channel::Publish(signal_publisher_, begin_signal);

  std::this_thread::sleep_for(std::chrono::seconds(1));

  // publish topic
  std::vector<std::future<void>> future_vec;
  std::vector<uint32_t> last_send_counts = GenerateLastSendCounts(plan.msg_count, plan.parallel_number);

  for (size_t ii = 0; ii < plan.parallel_number; ++ii) {
    auto publish_executor = executor_vec_[ii];
    auto publisher = publisher_vec_[0];

    std::promise<void> task_promise;
    future_vec.emplace_back(task_promise.get_future());
    publish_executor.Execute(
        [this, publisher, plan, last_send_counts, ii, task_promise{std::move(task_promise)}]() mutable {
          aimrt::protocols::example::BenchmarkMessage msg;
          msg.set_data(GenerateRandomString(plan.msg_size));

          uint32_t send_count = 0;

          uint32_t sleep_ns = static_cast<uint32_t>(1000000000 / plan.channel_frq);
          auto cur_tp = std::chrono::steady_clock::now();

          uint32_t messages_to_send = last_send_counts[ii + 1] - last_send_counts[ii];
          for (; send_count < messages_to_send; ++send_count) {
            if (!run_flag_.load()) [[unlikely]]
              break;

            msg.set_seq(last_send_counts[ii] + send_count);
            msg.set_timestamp(aimrt::common::util::GetCurTimestampNs());

            aimrt::channel::Publish(publisher, msg);

            cur_tp += std::chrono::nanoseconds(sleep_ns);
            std::this_thread::sleep_until(cur_tp);
          }

          task_promise.set_value();
        });
  }

  for (auto& fu : future_vec) fu.wait();

  std::this_thread::sleep_for(std::chrono::seconds(1));

  // publish end signal
  aimrt::protocols::example::BenchmarkSignal end_signal;
  end_signal.set_status(aimrt::protocols::example::BenchmarkStatus::End);
  end_signal.set_bench_plan_id(plan_id);

  AIMRT_INFO("Publish benchmark end signal, data: {}", aimrt::Pb2CompactJson(end_signal));
  aimrt::channel::Publish(signal_publisher_, end_signal);
}

}  // namespace aimrt::examples::cpp::pb_chn::benchmark_publisher_module